src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,153 @@
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        if (index == size - 1) {
+            return tail.item;
+        }
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (!isValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        Node<T> current = getNodeByIndex(index);
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is not valid");
+        }
+        Node<T> current = getNodeByIndex(index);
+        if (current.prev == null) {
+            head = current.next;
+            current.next = null;
+        } else if (current.next == null) {
+            tail = current.prev;
+            current.prev.next = null;
+            current.prev = null;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int nodeIndex = getIndexByValue(object);
+        if (nodeIndex == -1) {
+            return false;
+        } else {
+            remove(nodeIndex);
+            return true;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isValidIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private void addFirst(T value) {
+        head = tail = new Node<>(null, value, null);
+    }
+
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        tail = tail.next = newNode;
+    }
+
+    private int getIndexByValue(T value) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if ((value == null && current.item == null)
+                    || value != null && value.equals(current.item)) {
+                return i;
+            }
+            current = current.next;
+        }
+        return -1;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
```

Comment:
https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md#dont-iterate-your-nodes-from-the-head-to-the-tail-or-vice-versa

Please implement logic that will start iterating from head or tail depending on what is closer to target index